File: c#/libComputeDuck/ConstantFolder.cs
using System.Collections.Generic;

namespace ComputeDuck
{
    public class ConstantFolder
    {
        public List<Stmt> Fold(List<Stmt> stmts)
        {
            for (int i = 0; i < stmts.Count; ++i)
                stmts[i] = FoldStmt(stmts[i]);
            return stmts;
        }

        Stmt FoldStmt(Stmt stmt)
        {
            switch (stmt.type)
            {
                case AstType.RETURN:
                    return FoldReturnStmt((ReturnStmt)stmt);
                case AstType.EXPR:
                    return FoldExprStmt((ExprStmt)stmt);
                case AstType.SCOPE:
                    return FoldScopeStmt((ScopeStmt)stmt);
                case AstType.IF:
                    return FoldIfStmt((IfStmt)stmt);
                case AstType.WHILE:
                    return FoldWhileStmt((WhileStmt)stmt);
                case AstType.STRUCT:
                    return FoldStructStmt((StructStmt)stmt);
                default:
                    return stmt;
            }
        }
        Stmt FoldExprStmt(ExprStmt stmt)
        {
            stmt.expr = FoldExpr(stmt.expr);
            return stmt;
        }
        Stmt FoldIfStmt(IfStmt stmt)
        {
            stmt.condition = FoldExpr(stmt.condition);
            stmt.thenBranch = FoldStmt(stmt.thenBranch);

            if (stmt.elseBranch != null)
                stmt.elseBranch = FoldStmt(stmt.elseBranch);

            if (stmt.condition.type == AstType.BOOL)
            {
                if (((BoolExpr)stmt.condition).value == true)
                    return stmt.thenBranch;
                else
                    return stmt.elseBranch;
            }
            return stmt;
        }
        Stmt FoldScopeStmt(ScopeStmt stmt)
        {
            for (int i = 0; i < stmt.stmts.Count; ++i)
                stmt.stmts[i] = FoldStmt(stmt.stmts[i]);
            return stmt;
        }
        Stmt FoldWhileStmt(WhileStmt stmt)
        {
            stmt.condition = FoldExpr(stmt.condition);
            stmt.body = FoldStmt(stmt.body);
            return stmt;
        }
        Stmt FoldReturnStmt(ReturnStmt stmt)
        {
            stmt.expr = FoldExpr(stmt.expr);
            return stmt;
        }
        Stmt FoldStructStmt(StructStmt stmt)
        {
            for (int i = 0; i < stmt.members.Count; ++i)
                stmt.members[i] = new KeyValuePair<IdentifierExpr, Expr?>(stmt.members[i].Key, FoldExpr(stmt.members[i].Value));
            return stmt;
        }

        Expr FoldExpr(Expr expr)
        {
            switch (expr.type)
            {
                case AstType.NUM:
                    return FoldNumExpr((NumExpr)expr);
                case AstType.STR:
                    return FoldStrExpr((StrExpr)expr);
                case AstType.BOOL:
                    return FoldBoolExpr((BoolExpr)expr);
                case AstType.NIL:
                    return FoldNilExpr((NilExpr)expr);
                case AstType.IDENTIFIER:
                    return FoldIdentifierExpr((IdentifierExpr)expr);
                case AstType.GROUP:
                    return FoldGroupExpr((GroupExpr)expr);
                case AstType.ARRAY:
                    return FoldArrayExpr((ArrayExpr)expr);
                case AstType.INDEX:
                    return FoldIndexExpr((IndexExpr)expr);
                case AstType.PREFIX:
                    return FoldPrefixExpr((PrefixExpr)expr);
                case AstType.INFIX:
                    return FoldInfixExpr((InfixExpr)expr);
                case AstType.FUNCTION_CALL:
                    return FoldFunctionCallExpr((FunctionCallExpr)expr);
                case AstType.STRUCT_CALL:
                    return FoldStructCallExpr((StructCallExpr)expr);
                case AstType.REF:
                    return FoldRefExpr((RefExpr)expr);
                case AstType.FUNCTION:
                    return FoldFunctionExpr((FunctionExpr)expr);
                case AstType.STRUCT:
                    return FoldStructExpr((StructExpr)expr);
                default:
                    return expr;
            }
        }
        Expr FoldInfixExpr(InfixExpr expr)
        {
            expr.left = FoldExpr(expr.left);
            expr.right = FoldExpr(expr.right);

            return ConstantFold(expr);
        }
        Expr FoldNumExpr(NumExpr expr)
        {
            return expr;
        }
        Expr FoldBoolExpr(BoolExpr expr)
        {
            return expr;
        }
        Expr FoldPrefixExpr(PrefixExpr expr)
        {
            expr.right = FoldExpr(expr.right);
            return ConstantFold(expr);
        }
        Expr FoldStrExpr(StrExpr expr)
        {
            return expr;
        }
        Expr FoldNilExpr(NilExpr expr)
        {
            return expr;
        }
        Expr FoldGroupExpr(GroupExpr expr)
        {
            return FoldExpr(expr.expr);
        }
        Expr FoldArrayExpr(ArrayExpr expr)
        {
            for (int i = 0; i < expr.elements.Count; ++i)
                expr.elements[i] = FoldExpr(expr.elements[i]);
            return expr;
        }
        Expr FoldIndexExpr(IndexExpr expr)
        {
            expr.ds = FoldExpr(expr.ds);
            expr.index = FoldExpr(expr.index);
            return expr;
        }
        Expr FoldIdentifierExpr(IdentifierExpr expr)
        {
            return expr;
        }
        Expr FoldFunctionExpr(FunctionExpr expr)
        {
            for (int i = 0; i < expr.parameters.Count; ++i)
                expr.parameters[i] = (IdentifierExpr)FoldIdentifierExpr(expr.parameters[i]);

            expr.body = (ScopeStmt)FoldScopeStmt(expr.body);
            return expr;
        }
        Expr FoldFunctionCallExpr(FunctionCallExpr expr)
        {
            expr.name = FoldExpr(expr.name);
            for (int i = 0; i < expr.arguments.Count; ++i)
                expr.arguments[i] = FoldExpr(expr.arguments[i]);
            return expr;
        }
        Expr FoldStructCallExpr(StructCallExpr expr)
        {
            expr.callee = FoldExpr(expr.callee);
            expr.callMember = FoldExpr(expr.callMember);
            return expr;
        }
        Expr FoldRefExpr(RefExpr expr)
        {
            expr.refExpr = FoldExpr(expr.refExpr);
            return expr;
        }
        Expr FoldStructExpr(StructExpr expr)
        {
            for (int i = 0; i < expr.memberPairs.Count; ++i)
                expr.memberPairs[i] = new KeyValuePair<IdentifierExpr, Expr?>(expr.memberPairs[i].Key, FoldExpr(expr.memberPairs[i].Value));
            return expr;
        }

        Expr ConstantFold(Expr expr)
        {
            if (expr.type == AstType.INFIX)
            {
                var infix = (InfixExpr)expr;
                if (infix.left.type == AstType.NUM && infix.right.type == AstType.NUM)
                {
                    Expr newExpr = null;
                    if (infix.op == "+")
                        newExpr = new NumExpr(((NumExpr)infix.left).value + ((NumExpr)infix.right).value);
                    else if (infix.op == "-")
                        newExpr = new NumExpr(((NumExpr)infix.left).value - ((NumExpr)infix.right).value);
                    else if (infix.op == "*")
                        newExpr = new NumExpr(((NumExpr)infix.left).value * ((NumExpr)infix.right).value);
                    else if (infix.op == "/")
                        newExpr = new NumExpr(((NumExpr)infix.left).value / ((NumExpr)infix.right).value);
                    else if (infix.op == "==")
                        newExpr = new BoolExpr(((NumExpr)infix.left).value == ((NumExpr)infix.right).value);
                    else if (infix.op == "!=")
                        newExpr = new BoolExpr(((NumExpr)infix.left).value != ((NumExpr)infix.right).value);
                    else if (infix.op == ">")
                        newExpr = new BoolExpr(((NumExpr)infix.left).value > ((NumExpr)infix.right).value);
                    else if (infix.op == ">=")
                        newExpr = new BoolExpr(((NumExpr)infix.left).value >= ((NumExpr)infix.right).value);
                    else if (infix.op == "<")
                        newExpr = new BoolExpr(((NumExpr)infix.left).value < ((NumExpr)infix.right).value);
                    else if (infix.op == "<=")
                        newExpr = new BoolExpr(((NumExpr)infix.left).value <= ((NumExpr)infix.right).value);
                    return newExpr;
                }
                else if (infix.left.type == AstType.STR && infix.right.type == AstType.STR)
                {
                    var strExpr = new StrExpr(((StrExpr)infix.left).value + ((StrExpr)infix.right).value);
                    return strExpr;
                }
            }
            else if (expr.type == AstType.PREFIX)
            {
                var prefix = (PrefixExpr)expr;
                if (prefix.right.type == AstType.NUM && prefix.op == "-")
                {
                    var numExpr = new NumExpr(-((NumExpr)prefix.right).value);
                    return numExpr;
                }
                else if (prefix.right.type == AstType.BOOL && prefix.op == "not")
                {
                    var boolExpr = new BoolExpr(!((BoolExpr)prefix.right).value);
                    return boolExpr;
                }
            }

            return expr;
        }
    }
}

File: otherImpl/python/ConstantFolder.py
from Ast import *


class ConstantFolder:

    def fold(self, stmts: list[Stmt]) -> None:
        for s in stmts:
            s = self.__fold_stmt(s)

    def __fold_stmt(self, stmt: Stmt) -> Stmt:
        if stmt.type == AstType.RETURN:
            return self.__fold_return_stmt(stmt)
        elif stmt.type == AstType.EXPR:
            return self.__fold_expr_stmt(stmt)
        elif stmt.type == AstType.SCOPE:
            return self.__fold_scope_stmt(stmt)
        elif stmt.type == AstType.IF:
            return self.__fold_if_stmt(stmt)
        elif stmt.type == AstType.WHILE:
            return self.__fold_while_stmt(stmt)
        elif stmt.type == AstType.STRUCT:
            return self.__fold_struct_stmt(stmt)

    def __fold_expr_stmt(self, stmt: ExprStmt) -> Stmt:
        stmt.expr = self.__fold_expr(stmt.expr)
        return stmt

    def __fold_if_stmt(self, stmt: IfStmt) -> Stmt:
        stmt.condition = self.__fold_expr(stmt.condition)
        stmt.thenBranch = self.__fold_stmt(stmt.thenBranch)
        if stmt.elseBranch:
            stmt.elseBranch = self.__fold_stmt(stmt.elseBranch)
        if stmt.condition.type == AstType.BOOL:
            if stmt.condition.value == True:
                return stmt.thenBranch
            else:
                return stmt.elseBranch
        return stmt

    def __fold_scope_stmt(self, stmt: ScopeStmt) -> Stmt:
        for i in range(0, len(stmt.stmts)):
            stmt.stmts[i] = self.__fold_stmt(stmt.stmts[i])
        return stmt

    def __fold_while_stmt(self, stmt: WhileStmt) -> Stmt:
        stmt.condition = self.__fold_expr(stmt.condition)
        stmt.body = self.__fold_stmt(stmt.body)
        return stmt

    def __fold_return_stmt(self, stmt: ReturnStmt) -> Stmt:
        stmt.expr = self.__fold_expr(stmt.expr)
        return stmt

    def __fold_struct_stmt(self, stmt: StructStmt) -> Stmt:
        for s in stmt.members:
            s = self.__fold_expr(s)
        return stmt

    def __fold_expr(self, expr: Expr) -> Expr:
        if expr.type == AstType.NUM:
            return self.__fold_num_expr(expr)
        elif expr.type == AstType.STR:
            return self.__fold_str_expr(expr)
        elif expr.type == AstType.BOOL:
            return self.__fold_bool_expr(expr)
        elif expr.type == AstType.NIL:
            return self.__fold_nil_expr(expr)
        elif expr.type == AstType.IDENTIFIER:
            return self.__fold_identifier_expr(expr)
        elif expr.type == AstType.GROUP:
            return self.__fold_group_expr(expr)
        elif expr.type == AstType.ARRAY:
            return self.__fold_array_expr(expr)
        elif expr.type == AstType.INDEX:
            return self.__fold_index_expr(expr)
        elif expr.type == AstType.PREFIX:
            return self.__fold_prefix_expr(expr)
        elif expr.type == AstType.INFIX:
            return self.__fold_infix_expr(expr)
        elif expr.type == AstType.FUNCTION_CALL:
            return self.__fold_function_call_expr(expr)
        elif expr.type == AstType.STRUCT_CALL:
            return self.__fold_struct_call_expr(expr)
        elif expr.type == AstType.REF:
            return self.__fold_ref_expr(expr)
        elif expr.type == AstType.FUNCTION:
            return self.__fold_function_expr(expr)
        elif expr.type == AstType.STRUCT:
            return self.__fold_struct_expr(expr)
        else:
            return expr

    def __fold_infix_expr(self, expr: InfixExpr) -> Expr:
        expr.left = self.__fold_expr(expr.left)
        expr.right = self.__fold_expr(expr.right)
        return self.__constant_fold(expr)

    def __fold_num_expr(self, expr: NumExpr) -> Expr:
        return expr

    def __fold_bool_expr(self, expr: BoolExpr) -> Expr:
        return expr

    def __fold_prefix_expr(self, expr: PrefixExpr) -> Expr:
        expr.right = self.__fold_expr(expr.right)
        return self.__constant_fold(expr)

    def __fold_str_expr(self, expr: StrExpr) -> Expr:
        return expr

    def __fold_nil_expr(self, expr: NilExpr) -> Expr:
        return expr

    def __fold_group_expr(self, expr: GroupExpr) -> Expr:
        return self.__fold_expr(expr.expr)

    def __fold_array_expr(self, expr: ArrayExpr) -> Expr:
        for e in expr.elements:
            e = self.__fold_expr(e)
        return expr

    def __fold_index_expr(self, expr: IndexExpr) -> Expr:
        expr.ds = self.__fold_expr(expr.ds)
        expr.index = self.__fold_expr(expr.index)
        return expr

    def __fold_identifier_expr(self, expr: IdentifierExpr) -> Expr:
        return expr

    def __fold_function_expr(self, expr: FunctionExpr) -> Expr:
        for i in range(0, len(expr.parameters)):
            expr.parameters[i] = self.__fold_identifier_expr(
                expr.parameters[i])
        expr.body = self.__fold_scope_stmt(expr.body)
        return expr

    def __fold_function_call_expr(self, expr: FunctionCallExpr) -> Expr:
        expr.name = self.__fold_expr(expr.name)
        for i in range(0, len(expr.arguments)):
            expr.arguments[i] = self.__fold_expr(expr.arguments[i])
        return expr

    def __fold_struct_call_expr(self, expr: StructCallExpr) -> Expr:
        expr.callee = self.__fold_expr(expr.callee)
        expr.callMember = self.__fold_expr(expr.callMember)
        return expr

    def __fold_ref_expr(self, expr: RefExpr) -> Expr:
        expr.refExpr = self.__fold_expr(expr.refExpr)
        return expr

    def __fold_struct_expr(self, expr: StructExpr) -> Expr:
        for k, v in expr.memberPairs.items():
            v = self.__fold_expr(v)
        return expr

    def __constant_fold(self, expr: Expr) -> Expr:
        if expr.type == AstType.INFIX:
            if expr.left.type == AstType.NUM and expr.right.type == AstType.NUM:
                newExpr = None
                if expr.op == "+":
                    newExpr = NumExpr(expr.left.value+expr.right.value)
                elif expr.op == "-":
                    newExpr = NumExpr(expr.left.value-expr.right.value)
                elif expr.op == "*":
                    newExpr = NumExpr(expr.left.value*expr.right.value)
                elif expr.op == "/":
                    newExpr = NumExpr(expr.left.value/expr.right.value)
                elif expr.op == "==":
                    newExpr = BoolExpr(expr.left.value == expr.right.value)
                elif expr.op == "!=":
                    newExpr = BoolExpr(expr.left.value != expr.right.value)
                elif expr.op == ">":
                    newExpr = BoolExpr(expr.left.value > expr.right.value)
                elif expr.op == "<":
                    newExpr = BoolExpr(expr.left.value < expr.right.value)
                elif expr.op == ">=":
                    newExpr = BoolExpr(expr.left.value >= expr.right.value)
                elif expr.op == "<=":
                    newExpr = BoolExpr(expr.left.value <= expr.right.value)
                return newExpr
            elif expr.left.type == AstType.STR and expr.right.type == AstType.STR:
                return StrExpr(expr.left.value+expr.right.value)
        elif expr.type == AstType.PREFIX:
            if expr.right.type == AstType.NUM and expr.op == "-":
                return NumExpr(-expr.right.value)
            elif expr.right.type == AstType.BOOL and expr.op == "!":
                return BoolExpr(not (expr.right.value))
        return expr
